### PR TITLE
Add m2ctx tool

### PR DESCRIPTION
**Pull request by abaresk**
This is useful for generating context files for [decomp.me](https://decomp.me).

### CODE REVIEW

File: tools/m2ctx/m2ctx.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+root_dir = script_dir + "/../../"
+src_dir = root_dir + "src/"
+
+PREDEFINED_MACROS = [
+    "__STDC__", "__STDC_VERSION__", "__STDC_VERSION__", "__STDC_HOSTED__"]
+
+def get_c_dir(dirname):
+    for root, dirs, files in os.walk(src_dir):
+        for directory in dirs:
+            if directory == dirname:
```
Comment (PikalaxALT) [inline tools/m2ctx/m2ctx.py:20]:
```suggestion
        if dirname in dirs:
```

File: tools/m2ctx/m2ctx.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                return os.path.join(root, directory)
+
+
+def get_c_file(directory):
+    for root, dirs, files in os.walk(directory):
```
Comment (PikalaxALT) [inline tools/m2ctx/m2ctx.py:25]: might be faster to use the `glob` module?

File: tools/m2ctx/m2ctx.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    in_file = os.path.relpath(in_file, root_dir)
+    tmp_file = tempfile.NamedTemporaryFile(
+        prefix="c-includes-", suffix=".c", delete=False)
+
+    cmd = ["grep", "^#include ", in_file]
+    output = subprocess.check_output(cmd, cwd=root_dir, encoding="utf-8")
+
+    outfile = open(tmp_file.name, 'wt')
+    outfile.write(output)
+    outfile.close()
```
Comment (PikalaxALT) [inline tools/m2ctx/m2ctx.py:42]:
```suggestion
    tmp_file.write(output)
```

**Review by PikalaxALT — COMMENTED** [review]
I personally would replace this with a bash script rather than python.

**Comment by abaresk** [comment]
Switched to a bash script

File: tools/m2ctx/m2ctx.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+DEFINES="-DHEARTGOLD -DGAME_REMASTER=0 -DENGLISH -DPM_KEEP_ASSERTS -DSDK_ARM9 -DSDK_CODE_ARM -DSDK_FINALROM"
+
+
+generate-ctx () {
+    grep "^#include " "$1" > $TMP_IMPORTS
+    $GCC $FLAGS $INCLUDES $DEFINES $TMP_IMPORTS > $TMP_DECLARES
+
+    # Remove any line containing a predefined macro. If not removed, mwccarm
+    # generates compiler warnings.
+    sed '/__STDC__\|__STDC_VERSION__\|__STDC_VERSION__\|__STDC_HOSTED__/d' $TMP_DECLARES > $OUT_FILE
```
Comment (PikalaxALT) [inline tools/m2ctx/m2ctx.sh:19]: No need to use temp files, just pipe

```suggestion
    grep "^#include " "$1" | $GCC $FLAGS $INCLUDES $DEFINES $TMP_IMPORTS -x c - | sed '/__STDC__\|__STDC_VERSION__\|__STDC_VERSION__\|__STDC_HOSTED__/d' > $OUT_FILE
```

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

**Review by PikalaxALT — APPROVED** [review]

**Pull request merged**